lib/mix/lib/mix/tasks/escript.install.ex

```diff
@@ -28,21 +39,150 @@ defmodule Mix.Tasks.Escript.Install do
 
   ## Command line options
 
-    * `--sha512` - checks the escript matches the given sha512 checksum
+    * `--sha512` - checks the escript matches the given sha512 checksum. Only
+      applies to installations via URL or local path.
 
     * `--force` - forces installation without a shell prompt; primarily
       intended for automation in build systems like make
 
+    * `--submodules` - fetch repository submodules before building escript from
+      git or github
+
   """
 
   @behaviour Mix.Local.Installer
 
   @escript_file_mode 0o555 # only read and execute permissions
 
-  @switches [force: :boolean, sha512: :string]
+  @switches [force: :boolean, sha512: :string, submodules: :boolean]
   @spec run(OptionParser.argv) :: boolean
   def run(argv) do
-    Mix.Local.Installer.install({__MODULE__, :escript}, argv, @switches)
+    {opts, args} = OptionParser.parse!(argv, strict: @switches)
+
+    case args do
+      ["git" | rest] when rest != [] ->
+        raise_if_sha512("git", opts)
+        install_from_git(rest, opts)
+      ["github" | rest] when rest != [] ->
+        raise_if_sha512("github", opts)
+        install_from_github(rest, opts)
+      ["hex" | rest] when rest != [] ->
+        raise_if_sha512("hex", opts)
+        install_from_hex(rest, opts)
+      _ ->
+        Mix.Local.Installer.install({__MODULE__, :escript}, argv, @switches)
+    end
+  end
+
+  defp raise_if_sha512(name, opts) do
+    if opts[:sha512] do
+      Mix.raise "--sha512 is not supported for escript.install from #{name}"
+    end
+  end
+
+  defp install_from_github([repo | rest], opts) do
+    url = "https://github.com/#{repo}.git"
+    install_from_git([url | rest], opts)
+  end
+
+  defp install_from_git([url], opts) do
+    install_from_git([url, "branch", "master"], opts)
+  end
+
+  defp install_from_git([url, ref_type, ref], opts) do
+    git_opts =
+      ref_to_config(ref_type, ref) ++
+      [git: url, submodules: opts[:submodules]]
+
+    fetch_build_and_install_escript("new escript", {:"new escript", git_opts}, opts)
+  end
+
+  defp install_from_git([_url | rest], _opts) do
+    Mix.raise "escript.install received invalid git checkout spec: #{Enum.join(rest, " ")}"
+  end
+
+  defp ref_to_config("branch", branch), do: [branch: branch]
+
+  defp ref_to_config("tag", tag), do: [tag: tag]
+
+  defp ref_to_config("ref", ref), do: [ref: ref]
+
+  defp ref_to_config(ref_type, _) do
+    Mix.raise "escript.install expected one of \"branch\", \"tag\", or \"ref\". Got: \"#{ref_type}\""
+  end
+
+  defp install_from_hex([package_name], opts) do
+    install_from_hex([package_name, ">= 0.0.0"], opts)
+  end
+
+  defp install_from_hex([package_name, version], opts) do
+    fetch_build_and_install_escript(package_name, {String.to_atom(package_name), version}, opts)
```

Comment:
You can conflict with children using the same name so I would suggest
keeping them separate anyway.

On Monday, 6 June 2016, Eric Entin <notifications@github.com> wrote:

> In lib/mix/lib/mix/tasks/escript.install.ex
> <https://github.com/elixir-lang/elixir/pull/4742#discussion_r65961086>:
>
> > +  defp ref_to_config("branch", branch), do: [branch: branch]
> > +
> > +  defp ref_to_config("tag", tag), do: [tag: tag]
> > +
> > +  defp ref_to_config("ref", ref), do: [ref: ref]
> > +
> > +  defp ref_to_config(ref_type, _) do
> > +    Mix.raise "escript.install expected one of \"branch\", \"tag\", or \"ref\". Got: \"#{ref_type}\""
> > +  end
> > +
> > +  defp install_from_hex([package_name], opts) do
> > +    install_from_hex([package_name, ">= 0.0.0"], opts)
> > +  end
> > +
> > +  defp install_from_hex([package_name, version], opts) do
> > +    fetch_build_and_install_escript(package_name, {String.to_atom(package_name), version}, opts)
>
> Just FYI I reverted the serve_this_test package so it would not pollute
> hex, but it was just a duplicate of serve_this with a different name.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/4742/files/564c64f99ceccdf1ce09c8b01f0ee9956ab611b9#r65961086>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AATV2nLl96YIuzQAG1COKi880mTh1vlmks5qJIE-gaJpZM4Irrwy>
> .
>


-- 
Eric Meadows-Jönsson